yt/data_objects/selection_objects/base_objects.py

```diff
@@ -0,0 +1,1431 @@
+        with self._field_lock():
+            # At this point, we assume that any fields that are necessary to
+            # *generate* a field are in fact already available to us.  Note
+            # that we do not make any assumption about whether or not the
+            # fields have a spatial requirement.  This will be checked inside
+            # _generate_field, at which point additional dependencies may
+            # actually be noted.
+            while any(f not in self.field_data for f in fields_to_generate):
+                field = fields_to_generate[index % len(fields_to_generate)]
+                index += 1
+                if field in self.field_data:
+                    continue
+                fi = self.ds._get_field_info(*field)
+                try:
+                    fd = self._generate_field(field)
+                    if hasattr(fd, "units"):
+                        fd.units.registry = self.ds.unit_registry
+                    if fd is None:
+                        raise RuntimeError
+                    if fi.units is None:
+                        # first time calling a field with units='auto', so we
+                        # infer the units from the units of the data we get back
+                        # from the field function and use these units for future
+                        # field accesses
+                        units = getattr(fd, "units", "")
+                        if units == "":
+                            dimensions = ytdims.dimensionless
+                        else:
+                            dimensions = units.dimensions
+                            units = str(
+                                units.get_base_equivalent(self.ds.unit_system.name)
+                            )
+                        if fi.dimensions != dimensions:
+                            raise YTDimensionalityError(fi.dimensions, dimensions)
+                        fi.units = units
+                        self.field_data[field] = self.ds.arr(fd, units)
+                        mylog.warning(
+                            "Field %s was added without specifying units, "
+                            "assuming units are %s",
+                            fi.name,
+                            units,
+                        )
+                    try:
+                        fd.convert_to_units(fi.units)
+                    except AttributeError:
+                        # If the field returns an ndarray, coerce to a
+                        # dimensionless YTArray and verify that field is
+                        # supposed to be unitless
+                        fd = self.ds.arr(fd, "")
+                        if fi.units != "":
+                            raise YTFieldUnitError(fi, fd.units)
+                    except UnitConversionError:
+                        raise YTFieldUnitError(fi, fd.units)
+                    except UnitParseError:
+                        raise YTFieldUnitParseError(fi)
```

Comment:
lost in between mergers
```suggestion
                    except UnitConversionError as e:
                        raise YTFieldUnitError(fi, fd.units) from e
                    except UnitParseError as e:
                        raise YTFieldUnitParseError(fi) from e
```